src/Microsoft.Health.Fhir.Core/Features/Definition/CompartmentDefinitionManager.cs

```diff
@@ -0,0 +1,189 @@
+﻿using System.Linq;
+﻿using EnsureThat;
+﻿using Hl7.Fhir.ElementModel;
+﻿using Hl7.Fhir.Serialization;
+﻿using Hl7.Fhir.Utility;
+﻿using Hl7.FhirPath;
+﻿using Microsoft.Health.Extensions.DependencyInjection;
+﻿using Microsoft.Health.Fhir.Core.Data;
+﻿using Microsoft.Health.Fhir.Core.Exceptions;
+﻿using Microsoft.Health.Fhir.Core.Features.Definition.BundleNavigators;
+﻿using Microsoft.Health.Fhir.Core.Features.Search.Converters;
+﻿using Microsoft.Health.Fhir.Core.Models;
+﻿using Newtonsoft.Json;
+﻿using CompartmentType = Microsoft.Health.Fhir.ValueSets.CompartmentType;
+
+﻿namespace Microsoft.Health.Fhir.Core.Features.Definition
+{
+    /// <summary>
+    /// Manager to access compartment definitions.
+    /// </summary>
+    public class CompartmentDefinitionManager : IStartable, ICompartmentDefinitionManager
+    {
+        private readonly IModelInfoProvider _modelInfoProvider;
+
+        // This data structure stores the lookup of compartmentsearchparams (in the hash set) by ResourceType and CompartmentType.
+        private Dictionary<string, Dictionary<CompartmentType, HashSet<string>>> _compartmentSearchParamsLookup;
+
+        public CompartmentDefinitionManager(IModelInfoProvider modelInfoProvider)
+        {
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
+
+            _modelInfoProvider = modelInfoProvider;
+        }
+
+        public static Dictionary<string, CompartmentType> ResourceTypeToCompartmentType { get; } = new Dictionary<string, CompartmentType>
+        {
+            { KnownResourceTypes.Device, CompartmentType.Device },
+            { KnownResourceTypes.Encounter, CompartmentType.Encounter },
+            { KnownResourceTypes.Patient, CompartmentType.Patient },
+            { KnownResourceTypes.Practitioner, CompartmentType.Practitioner },
+            { KnownResourceTypes.RelatedPerson, CompartmentType.RelatedPerson },
+        };
+
+        public void Start()
+        {
+            // The json file is a bundle compiled from the compartment definitions currently defined by HL7.
+            // The definitions are available at https://www.hl7.org/fhir/compartmentdefinition.html.
+            using Stream stream = _modelInfoProvider.OpenVersionedFileStream("compartment.json");
+            using TextReader reader = new StreamReader(stream);
+            using JsonReader jsonReader = new JsonTextReader(reader);
+            var bundle = new BundleWrapper(FhirJsonNode.Read(jsonReader).ToTypedElement(_modelInfoProvider.StructureDefinitionSummaryProvider));
+            Build(bundle);
+        }
+
+        public bool TryGetSearchParams(string resourceType, string compartment, out HashSet<string> searchParams)
+        {
+            if (_compartmentSearchParamsLookup.TryGetValue(resourceType, out Dictionary<CompartmentType, HashSet<string>> compartmentSearchParams)
+                && compartmentSearchParams.TryGetValue(Enum.Parse<CompartmentType>(compartment), out searchParams))
```

Comment:
Should the dictionary just just be string to avoid parsing of the string every time it's executed?